System tags sidebar section

PVince81 opened the pull request:
Implements system tags sidebar sections inside a new app "systemtags"
#### TODOs
- [x] basic classes
- [x] REQUIRES: server APIs: https://github.com/owncloud/core/issues/20258
- [x] wire up model/collection to the backend APIs
- [x] write JS unit tests
#### Open questions:
- [x] should we read the tags from the file in an additional API call or through the file list PROPFIND ? (extra property "oc:systemtags"). Note that the tags are only displayed in the sidebar so we can save performance in the file list PROPFIND by loading them only later.
- [ ] rename "systemtags" app to "files_systemtags" since the section is limited to files ?
- [x] move "SystemTagCollection" and "SystemTagModel" JS classes to core for reusability ?
- [x] package the select2 init stuff into a separate view that can be reused ?
#### Tests:
- [x] TEST: initially selected tags appear
- [x] TEST: can assign existing tags
- [x] TEST: can unassign existing tags
- [x] TEST: can create new tags, tag gets created in the backend/db

@DeepDiver1975 @nickvergessen @blizzz FYI

Commit:
Added system tags GUI in sidebar

Added files details sidebar panel to assign/unassign/rename/delete
system tags.

nickvergessen commented:
> should we read the tags from the file in an additional API call or through the file list PROPFIND ? (extra property "oc:systemtags"). Note that the tags are only displayed in the sidebar so we can save performance in the file list PROPFIND by loading them only later.

I'd say make it possible to have it on propfind, but don't do it by default?

PVince81 commented:
I have cherry picked the backbone adapter from https://github.com/owncloud/core/pull/20882.
Well, actually I adjusted the adapter to make it work properly here.

So far, creating, assigning and unassigning tags work.

- [ ] BUG: cannot create several tags in a row, select2 doesn't show it properly?!
- [ ] TODO: check for race conditions, when a tag is being created, select2 already added it into its internal list (with a dummy id)

- [ ] TEST: create new tag and delete directly
- [ ] TEST: different visibilities ? (but probably something for later)

PVince81 commented:
- [ ] TODO: error handling
    - [ ] creating a tag that already exists (409)
    - [ ] assigning a tag already assigned concurrently (409)

PVince81 commented:
- [ ] BUG: select2 doesn't filter by whatever was typed in. Looks like it might not be possible to use both "create new item" and also "filter" at the same time ?!

PVince81 commented:
- [ ] TODO: use select2's "locked" attribute for value that have userAssignable: false
- [ ] TODO: filter out "userVisible: false"

nickvergessen commented:
> filter out "userVisible: false"

They should not be returned, to not leak the info?

PVince81 commented:
Yeah, I need to change the backend to not return them.

PVince81 commented:
Hacking select2, part one:
![select2-hack1](https://cloud.githubusercontent.com/assets/277525/12294855/8f0e11e4-b9fd-11e5-90ed-ea1471be8862.png)

Not yet close to the desired result, needs more work.

CC @jancborchardt

PVince81 commented:
- [ ] BUG: arrow keys skip already selected items
- [ ] TODO: try sorting in a way that has already selected at the top

PVince81 commented:
@jancborchardt can you confirm that this screenshot https://github.com/owncloud/core/pull/20847#issuecomment-171282748 is the direction we want to go before I dive into more dirty hacks to make it work ?

I assume that we wanted to have the search field inside the dropdown and not as part of the input field ? (a mock up would be nice)

Problems so far:
- [ ] cannot catch events inside the select2 result (pencil/trash icon), possible hack: http://stackoverflow.com/a/25849616
- [ ] the search field is embedded inside the input field, and not the results box

PVince81 commented:
- Fixed filtering.
- Added deletion when clicking on the trash icon.
- Added handler for rename (with ugly native prompt dialog), doesn't work yet => need PROPATCH support

- [ ] REQUIRES: PROPPATCH support in davclient: https://github.com/evert/davclient.js/issues/9

PVince81 commented:
@jancborchardt you can already try out this branch to get a (somewhat clumsy) feeling of the interaction.

jancborchardt commented:
Yep, looks good already. Besides styling details of course the only functionality change we should make is move the delete action into the edit interaction. That is, clicking edit will switch to rename mode, and there will also be the delete button there.

(The only worry I have is that in this case the Delete button might be directly in the same spot as edit … any input @owncloud/designers?)

PVince81 commented:
@jancborchardt thanks for the feedback.

I assume that the rename field will have to appear inline and replace the text, like it's done in the "New" menu for specifying a folder name.

Another a bit disturbing part is that if you want to delete a specific tag, you need to first type it in the "search" field, which is actually part of the input field. And only then you can edit/delete.

jancborchardt commented:
Yup, rename has gotta be inline.

What's the disturbing part exactly about that second realization?

PVince81 commented:
> What's the disturbing part exactly about that second realization?

It's not very intuitive to have to input a new tag in the field to trigger the search to be able to rename it, and then need to manually remove it again. But technically I didn't find a way with select2 to move the search field to be inside the dropdown instead of embedded in the input box.

PVince81 commented:
Here we go, added the inline field:
![select2-inlinefield](https://cloud.githubusercontent.com/assets/277525/12323965/d3a69430-babf-11e5-84a9-e45d08ee4542.png)

PVince81 commented:
and here a version with zoom: 100% (have to zoom out when developing):
![select2-inlinefield2](https://cloud.githubusercontent.com/assets/277525/12323991/044316fe-bac0-11e5-9ffa-f8235f355909.png)

Commit:
Implement toggleselect extension for select2

To make it possible to toggle selected values inside the dropdown

PVince81 commented:
Rebased on top of the backbone commit to get the latest version of the webdav adapter: https://github.com/owncloud/core/pull/20882

PVince81 commented:
Rebased to include the merged backbone stuff.

PVince81 commented:
Discussed with @nickvergessen, will provide two reusable tag select fields:
- single select (select one tag)
- multi select (like in this PR)

This means moving most of the JS code to "core/js/systemtags" to make it reusable.

PVince81 commented:
Moved to core/js/systemtags and still works :smile:

MorrisJobke commented on apps/systemtags/appinfo/app.php:
```diff
@@ -0,0 +1,40 @@
+	function() {
+		// FIXME: no public API for these ?
+		\OC_Util::addVendorScript('select2/select2');
+		\OC_Util::addVendorStyle('select2/select2');
```
vendor_script() inside the template :see_no_evil:

PVince81 replied:
there is no template in this kind of use case, that's the problem...

PVince81 replied:
PHP templates are so last year

MorrisJobke replied:
Sure -> create a public interface ;)

nickvergessen replied:
@MorrisJobke this is for the rightside bar, which is JS only?

nickvergessen replied:
App validation fails :speak_no_evil: 
```
    line   27: OC_Util - Static method of private class must not be called
    line   28: OC_Util - Static method of private class must not be called

PVince81 replied:
Looks like I'll have to add those missing public APIs... last time I tried I was told to use the template :hear_no_evil:

PVince81 commented:
I'll need more time. I need to split the select box to be reusable for selecting tags in other contexts where no "object to tag" assignment is needed but just the list of tag ids.

PVince81 commented:
Added JS unit tests.

PVince81 commented:
- [ ] BUG: extra comma

PVince81 commented:
Not sure what to do with the extra comma.
If I remove it, then the input field for the next tag is too close and adding a margin would not help.

Maybe instead of having them comma separated we could simply switch back to the old style where each tag appears as a box. It also makes it easier to distinguish them from the other fields like sharing.

Please note that the input field isn't a real input field, it's actually the tags input from select2 and works block-wise.

![systemtags1](https://cloud.githubusercontent.com/assets/277525/12420795/f56147e8-bebe-11e5-83d2-358bfe232a22.png)

@jancborchardt please try it out

PVince81 commented:
Rebased/squashed onto master, and also fixed server side unit tests.

PVince81 commented:
Would be good to get this merged soon / this week.

Please review @blizzz @nickvergessen @icewind1991 @rullzer 

I'm aware that there might still be some small UX / styling issues which we can fix separately.

nickvergessen commented on apps/systemtags/appinfo/info.xml:
```diff
@@ -0,0 +1,18 @@
+	<description>System-wide tags user interface</description>
+	<licence>AGPL</licence>
+	<author>Vincent Petry</author>
+	<shipped>true</shipped>
```
Deprecated field available: shipped => true

Only core/shipped.json is relevant

nickvergessen commented on apps/systemtags/appinfo/info.xml:
```diff
@@ -0,0 +1,18 @@
+	<licence>AGPL</licence>
+	<author>Vincent Petry</author>
+	<shipped>true</shipped>
+	<standalone/>
```
Deprecated field available: standalone

nickvergessen commented on apps/systemtags/appinfo/info.xml:
```diff
@@ -0,0 +1,18 @@
+	<default_enable/>
+	<version>0.1</version>
+	<dependencies>
+		<owncloud min-version="9.0" />
```
ownCloud max version requirement missing (will be an error in ownCloud 11 and later)

rullzer commented:
I get 

```
	Sabre\DAV\Exception\UnsupportedMediaType: HTTP/1.1 415 Unsupported Media Type

    /home/roeland/oc/core/apps/dav/lib/systemtag/systemtagplugin.php - line 116: OCA\DAV\SystemTag\SystemTagPlugin->createTag('{"name" "bar","...', 'application/jso...')
    [internal function] OCA\DAV\SystemTag\SystemTagPlugin->httpPost(Object(Sabre\HTTP\Request), Object(Sabre\HTTP\Response))
    /home/roeland/oc/core/3rdparty/sabre/event/lib/EventEmitterTrait.php - line 105: call_user_func_array(Array, Array)
    /home/roeland/oc/core/3rdparty/sabre/dav/lib/DAV/Server.php - line 459: Sabre\Event\EventEmitter->emit('method POST', Array)
    /home/roeland/oc/core/3rdparty/sabre/dav/lib/DAV/Server.php - line 248: Sabre\DAV\Server->invokeMethod(Object(Sabre\HTTP\Request), Object(Sabre\HTTP\Response))
    /home/roeland/oc/core/apps/dav/lib/server.php - line 118: Sabre\DAV\Server->exec()
    /home/roeland/oc/core/apps/dav/appinfo/v2/remote.php - line 29: OCA\DAV\Server->exec()
    /home/roeland/oc/core/remote.php - line 137: require_once('/home/roeland/o...')
    {main}
```

When I try to create a tag via the sidebar...

rullzer commented:
Seems to be since apparently firefox sets: `Content-Type: application/json; charset=UTF-8` which does not match in https://github.com/owncloud/core/blob/master/apps/dav/lib/systemtag/systemtagplugin.php#L150

PVince81 commented:
- [ ] BUG: firefox not happy to create tags

Commit:
Fix content type parsing in system tags plugin

rullzer commented:
[ ] UI glitch. Have multiple tags available. Select the top one. Everything si fine. Select the bottom one and the focus shifts to the top.

Not a deal breaker but still ;)

rullzer commented:
- [ ] Easy to createa tag conflicts.

1. userA logs in
2. userB logs in
3. userA creates a tag `holliday` for some pics
4. userB wants to use the same tag.
5. conflict

Probably won't happen often but erroring out here is not ideal.

rullzer commented:
- [ ] Tag list dropdown is only reordered on tag selection. Not deselection.

nickvergessen commented:
- [ ] UI glitch: Delete a tag, close the select2, tooltip for "Delete" is still there

nickvergessen commented:
:+1:

rullzer commented:
- [ ] Name of tag not updated

1. Assign a tag to a file
2. Unassign the tag
3. Rename the tag
4. Assign the tag agin

The old name shows up in the list. Not the new name

nickvergessen commented:
- [ ] Fatal: provisioning api test needs to be adjusted (man this test is nasty)

rullzer commented:
UI glitches can be fixes in different PR.

:+1:

PVince81 commented:
Note: I won't fix UI/UX glitches until the design is approved/tweaked by @jancborchardt 
Ideally I'd prefer having little boxes for tags:
- to be closer to select2's intended usage/UX
- to get rid of the comma problem
- it looks nicer

PVince81 commented:
I'll fix the CI issues

Commit:
Fix app metadata/test issues for system tags app

PVince81 commented:
Something weird with the smashbox job ?

nickvergessen commented on apps/systemtags/appinfo/info.xml:
```diff
@@ -0,0 +1,17 @@
+	<version>0.1</version>
+	<dependencies>
+		<owncloud min-version="9.0" />
+		<owncloud max-version="9.0" />
```
Can be on the same instance, or have to?

nickvergessen commented:
I think the smashbox is unrelated:
```
CRITICAL - recipients_1 - Exception occured: [Errno 2] No such file or directory: '/ssd/jenkins/workspace/smashbox-on-docker-ci@4/DOCKER_IMAGE/ubuntu_oc_lamp-git/TEST_NAME/test_sharePropagationGroups/mirallBranch/v2.0.2/slave/SMASH/smashdir/test_sharePropagationGroups/recipients_1/2/test/test2.txt
```

But when you push again, we can be sure

Commit:
Adjust system tags app max version

PVince81 commented:
@DeepDiver1975 please merge.

I'll move the open issues to the original ticket.

PVince81 commented:
I have listed the remaining issues here https://github.com/owncloud/core/issues/21815#issue-127709860
